testing/async/test_policy_overloads.h

```diff
@@ -0,0 +1,421 @@
+  //----------------------------------------------------------------------------
+  // For the specified postfix overload set, test the algorithm with several
+  // different policy configurations.
+  template <std::size_t postfix_idx>
+  static void run_basic_policy_tests(std::size_t const num_values)
+  {
+    // When a policy uses the default stream, the algorithm implementation
+    // should spawn a new stream in the returned event:
+    auto using_default_stream = [](auto& e) {
+      ASSERT_NOT_EQUAL(thrust::cuda_cub::default_stream(),
+                       e.stream().native_handle());
+    };
+
+    // When a policy uses a non-default stream, the implementation should pass
+    // the stream through to the output:
+    thrust::system::cuda::detail::unique_stream test_stream{};
+    auto using_test_stream = [&test_stream](auto& e) {
+      ASSERT_EQUAL(test_stream.native_handle(), e.stream().native_handle());
+    };
+
+    // Test the different types of policies:
+    run_basic_policy_test<postfix_idx>("(no policy)",
+                                       std::make_tuple(),
+                                       using_default_stream,
+                                       num_values);
+
+    run_basic_policy_test<postfix_idx>("thrust::device",
+                                       std::make_tuple(thrust::device),
+                                       using_default_stream,
+                                       num_values);
+
+    run_basic_policy_test<postfix_idx>(
+      "thrust::device(thrust::device_allocator<void>{})",
+      std::make_tuple(thrust::device(thrust::device_allocator<void>{})),
+      using_default_stream,
+      num_values);
+
+    run_basic_policy_test<postfix_idx>("thrust::device.on(test_stream.get())",
+                                       std::make_tuple(
+                                         thrust::device.on(test_stream.get())),
+                                       using_test_stream,
+                                       num_values);
+
+    run_basic_policy_test<postfix_idx>(
+      "thrust::device(thrust::device_allocator<void>{}).on(test_stream.get())",
+      std::make_tuple(
+        thrust::device(thrust::device_allocator<void>{}).on(test_stream.get())),
+      using_test_stream,
+      num_values);
+  }
+
+  // Wrap the basic_policy_test in a try/catch block to inject helpful info into
+  // any thrown unit test exceptions.
+  template <std::size_t postfix_idx, typename PolicyTuple, typename ValidateEvent>
+  static void run_basic_policy_test(std::string const &policy_desc,
+                                    PolicyTuple &&policy_tuple,
+                                    ValidateEvent &&validate,
+                                    std::size_t const num_values)
+  {
+    try
+    {
+      basic_policy_test<postfix_idx>(policy_tuple,
+                                     validate,
+                                     num_values);
+    }
+    catch (unittest::UnitTestException &exc)
```

Comment:
We have a bit of a bug when catching exceptions from non-CUDA backends, notably TBB and OpenMP; they wrap exceptions that occur in their threads in a special exception type (`tbb::captured_exception`, etc) which means these catch blocks won't match.

Reply:
The async framework is CUDA-only for the foreseeable future. If it gets expanded to other backends we can address this at that time.